---
title: "VitaminDPaperCode"
author: "Ankan Choudhury"
date: "2023-04-21"
output:
  html_document: default
  pdf_document:
  documentclass: article
  classoption: landscape
  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
knitr::opts_chunk$set(fig.width=20, fig.height=15) 
```

## R Markdown

```{r}
#Figure 2(A)
library(tidyverse)
library(readxl)
library(ggpubr)
sample_metadata <- read_excel("sample-metadata.xlsx") #Contains the metadata
head(sample_metadata)

VitaminDChange<-sample_metadata%>% #Creating a data frame with the serum Vitamin D changes in each subject
  select(Subject,Cohort,VitD_pre=`Vit D_pre`,VitD_post=`Vit D_post`)%>%
  distinct()%>%
  mutate(VitD_change =VitD_post-VitD_pre,
         VitD_pcchange =VitD_change/VitD_pre*100,
         VitD_pre_cat=if_else(VitD_pre<20,"Deficient (<20 ng/ml)",if_else(VitD_pre<30,"Insufficient (20-30 ng/ml)","Sufficient (>30 ng/ml)")))
head(VitaminDChange)

ggplot(VitaminDChange,aes(y=Cohort,x=VitD_change,fill=Cohort))+
  geom_boxplot(color="black",linewidth=1,width=0.25)+
  geom_dotplot(binaxis="x",binwidth = 3,dotsize=1.5,stroke=1,color="black",method = "histodot")+
  guides()+
  labs(y="Cohort",x="Change in serum Vitamin D\nlevels over 12 weeks (ng/mL)")+
  theme(text = element_text(size = 25),
        panel.background = element_rect(fill="white",color="grey",size=1),
        panel.border = element_rect(colour = "black", fill=NA, size=1),
        panel.grid.major= element_blank(),
        axis.text=element_text(size=25,color="black")
  )
  
  

```
```{r}
#Figure 2(B)
library(Polychrome) #making  a vector of 50 random colors
set.seed(07181990)
P50 <- createPalette(50, c("#FF0000", "#00FF00", "#0000FF"), range = c(0, 100))
P50 <- sortByHue(P50)
P50 <- as.vector(t(matrix(P50, ncol=10)))
names(P50) <- NULL

Genus <- read_excel("level-6.xlsx", 
                    sheet = "Genus")%>%
  pivot_longer(-c("index","Sample_Name","Day","Subject","Cohort")) #getting the relative abundance % at level6/Genus level 
head(Genus)

Genus%>%
  group_by(Cohort,Day,name)%>% #summing up the %RA of each taxon by Cohort and Day and filtering out only those above 15% for display
  summarize(sum=sum(value))%>%
  filter(sum>15)%>%
  ggplot(aes(fill=reorder(name,desc(sum)),y=sum,x=factor(Day),width=0.75))+
  geom_bar(position="fill",stat="identity",color="black")+
  scale_y_continuous(labels = scales::percent_format())+
  labs(y="Relative Abundance",x="Day",fill="Genus",title="Changes at Genus Level")+
  guides(fill=guide_legend(ncol=2))+
  facet_grid(~Cohort)+
  scale_fill_manual(values=P50)+
  theme(
    panel.background = element_rect(fill="white"),
    text = element_text(size=25,color="black"),
    panel.border = element_rect(size=1,fill="NA"),
    axis.text=element_text(size=15,color="black"),
    legend.text = element_text(size=15,color="black"),
    legend.title=element_text(size=25,face="bold",color="black"),
    legend.key.size = unit(1, 'cm'))

```
```{r}
#Figure 3A-D
AlphaVDMT <- read_excel("AlphaVDMT.xlsx",sheet = "Sheet1")# getting the alpha diversity indices generated by QIIME2
head(AlphaVDMT)

compare=list(c("Placebo","Treatment"))
compare2=list(c("1","7"),c("7","14"),c("14","78"),c("1","78"))

#Shannon vs Cohort faceted by day
ggplot(data=AlphaVDMT,aes(x=Cohort,y=shannon))+
  geom_boxplot(width=0.35,size=1.5,fill="grey")+
  geom_point(aes(group=Subject,fill=VitD_pre_cat),shape=21,color="black",size=5)+
  labs(x="Cohort",fill="Vitamin D level (Baseline)",y="Shannon Diversity")+
  stat_compare_means(comparisons=compare,method="t.test",size=8)+
  guides(color="none")+
  facet_grid(~factor(Day))+
  theme(
    panel.background = element_rect(fill="white"),
    text = element_text(size=30,color="black"),
    panel.border = element_rect(size=2,fill="NA"),
  )

#FaithPD vs Cohort faceted by day
ggplot(data=AlphaVDMT,aes(x=Cohort,y=faith))+
  geom_boxplot(width=0.35,size=1.5,fill="grey")+
  geom_point(aes(group=Subject,fill=VitD_pre_cat),shape=21,color="black",size=5)+
  labs(x="Cohort",fill="Vitamin D level (Baseline)",y="Faith Phylogenetic Diversity")+
  stat_compare_means(comparisons=compare,method="t.test",size=8)+
  guides(color="none")+
  facet_grid(~factor(Day))+
  theme(
    panel.background = element_rect(fill="white"),
    text = element_text(size=30,color="black"),
    panel.border = element_rect(size=2,fill="NA"),
  )

#Observed OTUs vs Cohort faceted by day
ggplot(data=AlphaVDMT,aes(x=Cohort,y=OTU))+
  geom_boxplot(width=0.35,size=1.5,fill="grey")+
  geom_point(aes(group=Subject,fill=VitD_pre_cat),shape=21,color="black",size=5)+
  labs(x="Cohort",fill="Vitamin D level (Baseline)",y="Observed OTUs")+
  stat_compare_means(comparisons=compare,method="t.test",size=8)+
  guides(color="none")+
  facet_grid(~factor(Day))+
  theme(
    panel.background = element_rect(fill="white"),
    text = element_text(size=30,color="black"),
    panel.border = element_rect(size=2,fill="NA"),
  )

#Shannon vs Day faceted by Cohort
ggplot(data=AlphaVDMT,aes(x=factor(Day),y=shannon))+
  geom_boxplot(width=0.35,size=1.5,fill="grey")+
  geom_point(aes(group=Subject,fill=VitD_pre_cat),shape=21,color="black",size=5)+
  stat_summary(fun=mean,geom="line",group=1,color="black",size=2)+
  labs(x="Day",fill="Vitamin D level (Baseline)",y="Shannon Diversity")+
  stat_compare_means(comparisons=compare2,method="t.test",paired=TRUE,size=8)+
  guides(fill="none")+
  facet_grid(~Cohort)+
  theme(
    panel.background = element_rect(fill="white"),
    text = element_text(size=30,color="black"),
    panel.border = element_rect(size=2,fill="NA"),
    )

#Faith PD vs Day faceted by Cohort
ggplot(data=AlphaVDMT,aes(x=factor(Day),y=faith))+
  geom_boxplot(width=0.35,size=1.5,fill="grey")+
  geom_point(aes(group=Subject,fill=VitD_pre_cat),shape=21,color="black",size=5)+
  stat_summary(fun=mean,geom="line",group=1,color="black",size=2)+
  labs(x="Day",fill="Vitamin D level (Baseline)",y="Faith Phylogentic Diversity")+
  stat_compare_means(comparisons=compare2,method="t.test",paired=TRUE,size=8)+
  guides(fill="none")+
  facet_grid(~Cohort)+
  theme(
    panel.background = element_rect(fill="white"),
    text = element_text(size=30,color="black"),
    panel.border = element_rect(size=2,fill="NA"),
  )

#Observed OTUs vs Day faceted by Cohort
ggplot(data=AlphaVDMT,aes(x=factor(Day),y=OTU))+
  geom_boxplot(width=0.35,size=1.5,fill="grey")+
  geom_point(aes(group=Subject,fill=VitD_pre_cat),shape=21,color="black",size=5)+
  stat_summary(fun=mean,geom="line",group=1,color="black",size=2)+
  labs(x="Day",fill="Vitamin D level (Baseline)",y="Observed OTUs")+
  stat_compare_means(comparisons=compare2,method="t.test",paired=TRUE,size=8)+
  guides(fill="none")+
  facet_grid(~Cohort)+
  theme(
    panel.background = element_rect(fill="white"),
    text = element_text(size=30,color="black"),
    panel.border = element_rect(size=2,fill="NA"),
  )


#Alternative figures with Alpha diversity indices calculated straight from QIIME2 derived feature table and phylogenetic tree

#Importing rooted phylogenetic tree created by QIIME2
tree<-ape::as.phylo(ape::read.tree("tree.nwk"))

#importing the feature table created by DADA2 in QIIME2
library(biomformat)
reads<-read_biom("feature-table.biom")
view(t(as.matrix(biom_data(reads))))

#Converting the feature table into a dataframe with samples as rows
OTUtable<-as.data.frame(t(as.matrix(biom_data(reads))))%>%
  rownames_to_column("Sample")%>%
  pivot_longer(-Sample)
head(OTUtable)

#Summarizing the number reads in each sample as a dataframe and finding the sample with the least number of reads for scaling/normalizing and rarefying purpose (minimum 500)
OTUcount<-OTUtable%>%
  group_by(Sample)%>%
  summarise(n=sum(value))%>%
  arrange(n)
head(OTUcount)

#Converting OTUtable dataframe into a count matrix as inputs for the distance matrix calculations
OTUmatrix<-OTUtable%>%
  pivot_wider(names_from = "name",values_from="value")%>%
  column_to_rownames("Sample")%>%
  as.matrix

library(picante)
#calculating the alpha diversity indices using the OTUs derived from the feature table and the phylogenetic tree
richness<-specnumber(OTUmatrix)%>% #Calculating richness/Observed OTU
  as.data.frame()%>%
  rownames_to_column("Sample")%>%
  rename_with(~"OTU",2)
head(richness)

shannon<-diversity(OTUmatrix)%>% #Calculating Shannon Diversity
  as.data.frame()%>%
  rownames_to_column("Sample")%>%
  rename_with(~"shannon",2)
head(shannon)

faith<-pd(OTUmatrix,tree)%>% #Calculating Faith phylogenetic Diversity
  rownames_to_column("Sample")%>%
  rename_with(~"faithPD",2)
head(faith)

#joining all the indices together and merging with the metadata and the Vitamin D category variable 
AlphaVDMT2<-inner_join(richness,shannon,by="Sample")%>% 
  inner_join(faith,by="Sample")%>%
  inner_join(OTUcount,by="Sample")%>%
  select(-SR)%>%
  pivot_longer(-c(Sample,n))%>%
  rename_with(~"Alpha",3)%>%
  inner_join(sample_metadata,by="Sample")%>%
  inner_join(VitaminDChange[,c("Subject","VitD_pre_cat")],by="Subject")
head(AlphaVDMT2)

alphaList<-c("Shannon Diversity","Faith Phylogenetic\nDiversity","Observed OTUs")
names(alphaList)<-c("shannon","faithPD","OTU")

#Alpha indices vs Day faceted by Cohort
ggplot(AlphaVDMT2,aes(x=factor(Day),y=value))+
  geom_boxplot(width=0.35,size=1.5,fill="grey")+
  geom_point(aes(group=Subject,fill=VitD_pre_cat),shape=21,color="black",size=5)+
  stat_compare_means(comparisons=compare2,method="wilcox.test",paired=TRUE,size=4,bracket.size = 1,step.increase = 0.05)+
  stat_summary(group=1,fun=mean,geom="line",color="black",size=2)+
  facet_grid(Alpha~Cohort,scales="free_y",labeller=labeller(Alpha=alphaList))+
  theme_classic()+
  labs(x="Day")+
  ylab(NULL)+
  guides(fill="none")+
  theme(
    text=element_text(size=25,face="bold"),
    panel.border = element_rect(linewidth=1,color="black",fill=NA),
    strip.background = element_rect(fill="grey"),
    strip.background.y=element_blank(),
    strip.placement = "outside")

#Alpha indices vs Cohort faceted by Day
ggplot(AlphaVDMT2,aes(x=Cohort,y=value))+
  geom_boxplot(width=0.35,size=1.5,fill="grey")+
  geom_point(aes(group=Subject,fill=VitD_pre_cat),shape=21,color="black",size=5)+
  stat_compare_means(comparisons=compare,method="wilcox.test",size=4,bracket.size = 1,step.increase = 0.05)+
  facet_grid(Alpha~factor(Day),scales="free_y",labeller=labeller(Alpha=alphaList))+
  theme_classic()+
  labs(x="Day")+
  ylab(NULL)+
  theme(
    text=element_text(size=25,face="bold"),
    panel.border = element_rect(linewidth=1,color="black",fill=NA),
    strip.background = element_rect(fill="grey"),
    strip.background.y=element_blank(),
    strip.placement = "outside")


```

```{r}
#Figure 4
library(vegan)
#Calculating Aitchison's Distance from the OTUMatrix and making a dataframe containing the Day and Cohort data of each samples
aitch_dist<-vegdist(OTUmatrix,method="aitchison",pseudocount=1)%>%
  as.matrix()
head(aitch_dist)

#Converting the matrix to data frame and pivoting it lengthwise
aitch_dist_df<-aitch_dist%>% 
  as.data.frame()%>%
  rownames_to_column("Sample")%>%
  pivot_longer(-Sample)%>%
  filter(name<Sample)%>%
  inner_join(sample_metadata[,c("Sample","Day","Cohort")],by="Sample")%>% #joining Subject, Day and Cohort for both the samples between which the Aitchison's distance was calculated
  merge(sample_metadata[,c("Sample","Day","Cohort")],by.x="name",by.y="Sample")

head(aitch_dist_df)

Cohorts<-c("Within Placebo group","Within Treatment group")
names(Cohorts)<-c("Placebo","Treatment")

#Boxplot by filtering the data to keep only the distances between the same group in each Day, Figure 4A&B
aitch_dist_df%>%
  filter(Day.x==Day.y&Cohort.x==Cohort.y)%>%
  ggplot(aes(x=factor(Day.x),y=value))+
  geom_boxplot(width=0.35,size=1.5,fill="grey")+
  geom_jitter(width=0.05)+
  stat_summary(fun=mean,geom="line",group=1,color="black",size=2)+
  labs(x="Day",y="Aitchison's Distance")+
  ylim(20,130)+
  stat_compare_means(comparisons=compare2,method="wilcox.test",size=7.5)+
  facet_grid(~Cohort.x,labeller=labeller(Cohort.x=Cohorts))+
  theme(
    panel.background = element_rect(fill="white"),
    text = element_text(size=30,color="black"),
    panel.border = element_rect(size=2,fill="NA"),
  )

#Figure 4C
library(ggrepel)

#Performing PCOA based on the aitchison's distances using CMD scale
aitchpcoa<-aitch_dist%>%
  cmdscale(k=2,eig=TRUE,add=TRUE)

#Extracting the top 2 axis explaining most of the variations based on the eigen values
aitchpcoaaxispc<-round(100*aitchpcoa$eig/sum(aitchpcoa$eig),2)
view(aitchpcoaaxispc[1:10])

#Converting the PCOA into a dataframe with Axis coordinates for each sample and obtaining Subject, DAy and Cohort data for each sample from the metadata
aitchpcoaplot<-as.data.frame(aitchpcoa$points)%>%
  rename_with(~c("Axis1","Axis2"),c(1,2))%>%
  rownames_to_column("Sample")%>%
  inner_join(sample_metadata[,c("Sample","Day","Cohort","Subject")],by="Sample")%>%
  mutate(index=substr(Sample,6,10))
head(aitchpcoaplot)

#Filtering the previous data frame to only contain samples from Days 7,14 and 78 for PERMANOVA analysis
aitch_permanova<-aitch_dist%>%
  as.data.frame()%>%
  rownames_to_column("Sample")%>%
  inner_join(aitchpcoaplot,by="Sample")%>%
  filter(Day>1)
head(aitch_permanova)

#Converting the previously obtained dataframe into a distance objcet and performing PERMANOVA analysis with Cohort as the factor variable
aitch_permanova_dist<-as.dist(select(aitch_permanova,all_of(aitch_permanova[["Sample"]])))
aitch_adonis<-adonis2(aitch_permanova_dist~Cohort,data=aitch_permanova,permutations=10000)

#Ploting the PCOA distribution of samples from Days 7,14 and 78 along with the PERMANOVA stats
aitchpcoaplot%>%
  filter(Day>1)%>%
  ggplot(aes(x=Axis1,y=Axis2,fill=Cohort))+
  geom_point(color="black",shape=21,stroke=1.5,alpha=0.75,size=7.5)+
  scale_fill_manual(values=c("#416DD4",  "#FFBC33"))+
  scale_color_manual(values=c("#416DD4",  "#FFBC33"))+
  labs(x=paste("PCoA Axis 1 (",aitchpcoaaxispc[1],"%)",sep=""),
       y=paste("PCoA Axis 2 (",aitchpcoaaxispc[2],"%)",sep=""),
       title="PCoA based on Aitchison Distance of samples from\n Days 7, 14 and 84")+
  geom_text_repel(aes(label=index),max.overlaps=200, hjust = "right",force=1,size=6)+
  stat_ellipse(aes(color=Cohort),type="t",size=1,linetype=2)+
  annotate(geom="text",x=20,y=25,label=paste("PERMANOVA\nF-statistics =",round(aitch_adonis["Cohort","F"],4),"\nFDR p-value =",round(aitch_adonis["Cohort","Pr(>F)"],4)),fontface="bold",size=7.5)+
  theme(text = element_text(size = 25,face="bold"),
        panel.background = element_rect(fill="white",color="grey",linewidth=2),
        panel.border = element_rect(colour = "black", fill=NA, size=2),
        panel.grid.major= element_blank(),
        axis.text=element_text(size=25,color="black"),
        legend.key.size = unit(10, "mm"))

library(tidyr)
#Figure 4D
#Plotting the centroids (mean and Median) of the axis coordinates of the samples form Day 7,14 and 78
aitchpcoaplot%>%
  filter(Day>1)%>%
  group_by(Subject)%>%
  summarize(median_Axis1=median(Axis1),
            median_Axis2=median(Axis2),
            mean_Axis1=mean(Axis1),
            mean_Axis2=mean(Axis2))%>%
  inner_join(VitaminDChange[,c("Cohort","Subject")],by="Subject")%>%
  pivot_longer(-c("Subject","Cohort"))%>%
  separate(name, sep = "_", into = c("Centroid", "Axes"))%>%
  pivot_wider(values_from="value",names_from="Axes")%>%
  mutate(index=substr(Subject,6,7))%>%
  ggplot(aes(x=Axis1,y=Axis2,fill=Centroid))+
  geom_point(color="black",shape=21,stroke=1.5,alpha=0.75,size=7.5)+
  labs(x=paste("PCoA Axis 1 (",aitchpcoaaxispc[1],"%)",sep=""),
       y=paste("PCoA Axis 2 (",aitchpcoaaxispc[2],"%)",sep=""),
       title="PCoA based on Aitchison Distance of samples from\n Days 7, 14 and 84")+
  geom_text_repel(aes(label=index),max.overlaps=200, hjust = "right",force=1,size=6)+
  stat_ellipse(aes(color=Centroid),type="t",size=1,linetype=2)+
  facet_grid(~Cohort)+
  theme(text = element_text(size = 25,face="bold"),
        panel.background = element_rect(fill="white",color="grey",linewidth=2),
        panel.border = element_rect(colour = "black", fill=NA, size=2),
        panel.grid.major= element_blank(),
        axis.text=element_text(size=25,color="black"),
        legend.key.size = unit(10, "mm"))
  
#Figure 4E
#Filtering the dataframe with the aitchison's distance to only contain distances between the same subject over all the days
aitch_dist_df2<-aitch_dist%>%
  as.data.frame()%>%
  rownames_to_column("Sample")%>%
  pivot_longer(-Sample)%>%
  filter(name<Sample)%>%
  inner_join(sample_metadata[,c("Sample","Day","Subject")],by="Sample")%>%
  merge(sample_metadata[,c("Sample","Day","Subject")],by.x="name",by.y="Sample")%>%
  filter(Subject.x==Subject.y,Day.x!=Day.y)

#Plotting the stability (inverse of aitchison's distance) for every subject against the % change in their vitamin D serum level  
VitaminDChange%>%
  inner_join(sample_metadata[,c("Sample","Subject")],by="Subject")%>%
  merge(aitch_dist_df2,by.x="Subject",by.y="Subject.x")%>%
  mutate(Stability=1/value)%>%
  filter(Cohort=="Treatment")%>%
  mutate(Change_class=ifelse(VitD_pcchange<60,"Below 60%",ifelse(VitD_pcchange<120,"60% to 120%","Above 120%")))%>%
  ggplot(aes(x=VitD_pcchange,y=Stability))+
  geom_point(aes(fill=Subject),shape=21,size=5,color="black",stroke=1.5)+
  geom_smooth(aes(group=Change_class),method="lm",formula=y ~ x,size=2,color="black",linetype=4)+
  labs(x="% Change in serum Vitamin D levels", y="Stability",title="Treatment subjects")+
  guides(fill="none",color="none")+
  facet_grid(~factor(Change_class,levels=c("Below 60%","60% to 120%","Above 120%")),scales="free_x")+
  stat_regline_equation(aes(group=Change_class),label.y=0.03,size=7.5,color="black")+
  stat_cor(aes(group=Change_class,label=..r.label..),label.y=0.02925,size=7.5,color="black")+
  stat_cor(aes(group=Change_class,label=..rr.label..),label.y=0.0285,size=7.5,color="black")+
  stat_cor(aes(group=Change_class,label=..p.label..),label.y=0.02775,size=7.5,color="black")+
  theme(
    panel.background = element_rect(fill="white"),
    text = element_text(size=30,color="black",face="bold"),
    panel.border = element_rect(size=2,fill="NA"),
  )

#Supplementary figure
VitaminDChange%>%
  inner_join(sample_metadata[,c("Sample","Subject")],by="Subject")%>%
  merge(aitch_dist_df2,by.x="Subject",by.y="Subject.x")%>%
  mutate(Stability=1/value)%>%
  filter(Cohort=="Placebo",Subject!="VDMT040")%>%
  mutate(Change_class=if_else(VitD_pcchange<(-10),"Below -10%",if_else(VitD_pcchange<(10),"-10% to 10%","Above 10%")))%>%
  ggplot(aes(x=VitD_pcchange,y=Stability))+
  geom_point(aes(fill=Subject),shape=21,size=5,color="black",stroke=1.5)+
  geom_smooth(aes(group=Change_class),method="lm",formula=y ~ x,size=2,color="black",linetype=4)+
  labs(x="% Change in serum Vitamin D levels", y="Stability",title="Placebo subjects")+
  guides(fill="none",color="none")+
  facet_grid(~factor(Change_class,levels=c("Below -10%","-10% to 10%","Above 10%")),scales="free_x")+
  stat_regline_equation(aes(group=Change_class),label.y=0.03,size=7.5,color="black")+
  stat_cor(aes(group=Change_class,label=..r.label..),label.y=0.02925,size=7.5,color="black")+
  stat_cor(aes(group=Change_class,label=..rr.label..),label.y=0.0285,size=7.5,color="black")+
  stat_cor(aes(group=Change_class,label=..p.label..),label.y=0.02775,size=7.5,color="black")+
  theme(
    panel.background = element_rect(fill="white"),
    text = element_text(size=30,color="black",face="bold"),
    panel.border = element_rect(size=2,fill="NA"),
  )

```
```{r}
#Figure 5 A-C
library(pheatmap)
#Getting the taxa with significantly different abundabce compared to Day 1 for the Placebo Group and their significance markers
PlaceboChangeOnlySigs <- read_excel("Top100taxaVDMT.xlsx", 
                            sheet = "PlaceboTaxaChangeOnlySigs")
PlaceboChangeOnlySigspval <- read_excel("Top100taxaVDMT.xlsx", 
                         sheet = "PlaceboTaxaChangeOnlySigspval2")
PlaceboChangeOnlySigs2<-as.matrix(PlaceboChangeOnlySigs[,c(2:30)])
rownames(PlaceboChangeOnlySigs2)<-PlaceboChangeOnlySigs$Day
PlaceboChangeOnlySigspval2<-PlaceboChangeOnlySigspval[,c(2:30)]
PlaceboChangeOnlySigspval2<-replace(PlaceboChangeOnlySigspval2,is.na(PlaceboChangeOnlySigspval2)," ")
rownames(PlaceboChangeOnlySigspval2)<-PlaceboChangeOnlySigspval$Day

#Plotting a heatmap with differentially abundant taxa
pheatmap(PlaceboChangeOnlySigs2,display_num=PlaceboChangeOnlySigspval2,
         color=colorRampPalette(c("#416DD4", "#565E6A", "#FFBC33"))(100),scale="none",
         cluster_rows = FALSE,show_rownames=TRUE,
         border_color=NA,
         fontsize_number=25,
         fontsize_row=25,
         fontsize_col= 25,
         number_color="white",
         cellheight=100,
         fontsize=20,
         main="Taxonomic abundance changes across Days in Placebo group")

#Getting the taxa with significantly different abundabce compared to Day 1 for the Treatment Group and their significance markers
TreatmentChangeOnlySigs <- read_excel("Top100taxaVDMT.xlsx", 
                                    sheet = "TreatmentTaxaChangeOnlySigs")
TreatmentChangeOnlySigspval <- read_excel("Top100taxaVDMT.xlsx", 
                                        sheet = "TreatmentTaxaChangeOnlySigpval2")
TreatmentChangeOnlySigs2<-as.matrix(TreatmentChangeOnlySigs[,c(2:34)])
rownames(TreatmentChangeOnlySigs2)<-TreatmentChangeOnlySigs$Day
TreatmentChangeOnlySigspval2<-TreatmentChangeOnlySigspval[,c(2:34)]
TreatmentChangeOnlySigspval2<-replace(TreatmentChangeOnlySigspval2,is.na(TreatmentChangeOnlySigspval2)," ")
rownames(TreatmentChangeOnlySigspval2)<-TreatmentChangeOnlySigspval$Day

#Plotting a heatmap with differentially abundant taxa
pheatmap(TreatmentChangeOnlySigs2,display_num=TreatmentChangeOnlySigspval2,
         color=colorRampPalette(c("#416DD4", "#565E6A", "#FFBC33"))(100),scale="none",
         cluster_rows = FALSE,show_rownames=TRUE,
         border_color=NA,
         fontsize_number=25,
         fontsize_row=25,
         fontsize_col= 20,
         number_color="white",
         cellheight=100,
         fontsize=20,
         main="Taxonomic abundance changes across Days in Treatment group")

#Getting the top 5 taxa that significantly chnaged in abundance compared to Day 1 for Treatment group
TreatmentTaxaChange <- read_excel("Top100taxaVDMT.xlsx",sheet = "TreatmentTaxaChange2")%>%
  mutate(Change2=ifelse(Change>0,"Increase","Decrease"))
ggplot(TreatmentTaxaChange,aes(x=reorder(taxa,Change),y=Change))+
  geom_bar(stat="identity",color="black",aes(fill=Change2))+
  facet_grid(~Day,scales="free_x")+
  labs(x="Genus",y="Change in Relative Abundance",title="Top 5 taxa that changed significantly in abundance in Treatment group",fill="")+
  theme(
    panel.background = element_rect(fill="white"),
    text = element_text(size=25,color="black"),
    panel.border = element_rect(size=2,fill="NA"),
    axis.text.x=element_text(angle=90,hjust=1,vjust=1,color="black")
  )

```


```{r}
#Regression analysis of individual taxa with VitaminD change and Figure 6A-C
library(broom)
GenusVitChange<-read_excel("level-6.xlsx",sheet = "Genus2")%>%              #getting abundance data of the taxa at level 6
  pivot_longer(-c("index","Sample_Name","Day","Subject","Cohort"))%>%
  inner_join(VitaminDChange[,c("Subject","VitD_pcchange")],by="Subject")%>% #joining the Taxa abundance with the Vitamin D changes
  group_by(Cohort,name)%>%
  mutate(meanAbundance=mean(value))%>%              #filtering only those taxa that has a mean abundance over 0.1% in each Cohort
  filter(meanAbundance>0.1)%>%
  mutate(ChangeCat=ifelse(VitD_pcchange>=50,">50%","<50%")) #creating a category variable that tells whether the change in Vitamin D is over 50% or not

Sig_taxa<-GenusVitChange %>%    #figuring out the taxa that changed their abundance significantly after Day 1 based on whether the subject had a Vitamin D level change greater than 50% or not maong the Treatment group
  filter(Cohort=="Treatment",Day>1)%>%
  nest(data=-name)%>%
  mutate(test=map(.x=data,~aov(value~ChangeCat,data=.x)%>%tidy))%>%
  unnest(test)%>%
  mutate(FDR=p.adjust(p.value,method="BH"))%>%
  filter(FDR<0.01)

#Plotting the abundance of those taxa as Figure 6A 
GenusVitChange %>%
  inner_join(Sig_taxa,by="name")%>%
  filter(Cohort=="Treatment"&Day>1)%>%
  ggplot(aes(x=value,y=ChangeCat,color=ChangeCat,fill=ChangeCat))+
  geom_boxplot(width=.35,color="black",size=2)+
  geom_jitter(shape=21,size=7.5,stroke=1.5,color="black",position=position_jitterdodge(dodge.width=1,jitter.width=0.15))+
  facet_wrap(~name)+
  xlab("log of Relative Abundance")+ylab("Change in Vitamin D")+
  guides(color="none",fill="none")+
  scale_x_log10()+
  theme_classic()+
  theme(text=element_text(size=20,color="black",face="bold"),
        strip.text=element_text(size=20,color="black",face="bold.italic"),
        axis.text = element_text(size=20,color="black",face="bold"))


#Regression analysis to see whether the abundance of these taxa on Day 1 are a significant predictor of the gain in Vitamin D level
GenusRegression<-read_excel("level-6.xlsx",sheet = "Genus2")%>% #getting the data
  inner_join(VitaminDChange[,c("Subject","VitD_pre","VitD_post","VitD_pcchange")],by="Subject")%>% #joining with Vitamin D change data and creating a catgory of change
  filter(Day==1&Cohort=="Treatment") %>%
  mutate(ChangeCat=ifelse(VitD_pcchange>=50,">50%","<50%"))

Regression<-lm(VitD_pcchange~VitD_pre+LachnospiraceaeCAG56+Dialister+Corynebacterium1+Intestinibacter,data=GenusRegression) #performing reg analysis with the 4 bacteria found previously
summary(Regression)

Regression<-lm(VitD_pcchange~VitD_pre+LachnospiraceaeCAG56+Dialister+Intestinibacter,data=GenusRegression) #performing reg analysis by dropping Corynebacterium1
summary(Regression)

library(car)
avPlots(Regression, col="red",pch=10,lwd=5,main="%Change in VitaminD against selected bacteria and baseline Vitamin D level") #regression plots with individual variables, only Vitamin D pre level has a signiifcant effect on the change of Vitamin D

#Random forest analysis and Figure 6B&C

library(randomForest)
library(ROCR)

RandomForest<-read_excel("level-6.xlsx",sheet = "Genus2")%>%              #getting abundance data of the taxa at level 6
  inner_join(VitaminDChange[,c("Subject","VitD_pcchange")],by="Subject")%>% #joining the Taxa abundance with the Vitamin D changes
  mutate(ChangeCat=ifelse(VitD_pcchange>=50,">50%","<50%"))%>%
  filter(Day>1&Cohort=="Treatment") %>%
  select(-c(1,2,3,4,5,446))

set.seed(07181990)
rf<-randomForest(as.factor(ChangeCat)~.,data=RandomForest,ntree=10000,type="classification",proximity=TRUE) #performing the random forest analysis

Imp<-as.data.frame(importance(rf))
Imp$variable<-rownames(Imp)
Imp<-Imp[order(Imp$MeanDecreaseGini,decreasing=TRUE),]%>%
  filter(MeanDecreaseGini>0.2)
Imp%>% #Figure 6B
  head(n=20)%>%
  ggplot(aes(x=MeanDecreaseGini*10,y=reorder(variable,MeanDecreaseGini)))+
  geom_bar(stat="identity",aes(fill=-MeanDecreaseGini*10),color="black")+
  labs(title="Top taxa contributing to the Random Forest model in\npredicting change of Vitamin D >50%",
       x="Importance",y="Taxa",fill="")+
  guides(fill="none")+
  theme_classic()+
  theme(text=element_text(size=25,face="bold"),
        axis.text.y = element_text(size=20,face="italic",color="black"),
        legend.key.size=unit(1,"cm"),
        axis.text.x=element_text(size=20,face="bold",color="black"),)

RandomForest<-RandomForest%>%
  select(ChangeCat,Imp$variable)
rf<-randomForest(as.factor(ChangeCat)~.,data=RandomForest,ntree=10000,type="classification",proximity=TRUE)
rf

pred1=predict(rf,type = "prob")
perf = prediction(pred1[,2], RandomForest$ChangeCat)
# 1. Area under curve
auc = performance(perf, "auc")
auc
# 2. True Positive and Negative Rate
pred3 = performance(perf, "tpr","fpr")
auc_ROCR <- performance(perf, measure = "auc")
auc_ROCR@y.values[[1]]
pred2<-as.data.frame(cbind(pred3@x.values[[1]],pred3@y.values[[1]]))
# 3. Plotting the ROC curve- Figure 6C
ggplot(pred2,aes(x=V1,y=V2))+
  geom_line(color="red",size=3)+
  geom_abline(slope=1,size=2,color="black",linetype=2)+
  geom_text(x=0.5,y=0.75,label=paste("AUCROC = ",round(auc_ROCR@y.values[[1]],3)),size=7.5,color="black")+
  labs(title="ROC of the Random Forest Model",x="False Positive rate",y="True Positive rate")+
  theme_classic()+
  theme(text=element_text(size=25,face="bold"),
        axis.text= element_text(size=20,face="bold",color="black"),
        panel.border=element_rect(linewidth=2,color="black",fill=NA))
  





```